### PR TITLE
Update dlpack to v1.0rc

### PR DESCRIPTION
**Pull request by ndgrigorian**
This PR bumps the version of dlpack by bringing in the new `dlpack.h` from v1.0rc.

As a result of changes introduced in v1.0rc
- `get_build_dlpack_version` now returns a tuple giving the major and minor DLPack version dpctl was built with
- support for the new `DLManagedTensorVersioned` struct has been added, including full support of the new `flags` attribute
- `__dlpack__` method of `usm_ndarray` has been updated to choose between `DLManagedTensorVersioned` and `DLManagedTensor` paths based on the new `max_version` keyword argument, and also now has `dl_device` and `copy` keyword arguments
- `from_dlpack` method has been updated with `device` and `copy` keyword arguments and now has a path for arrays supporting the new `__dlpack__` interface and `DLManagedTensorVersioned`. `from_dlpack` also honors the read-only bitmask that can now be present on a `DLManagedTensorVersioned` object
- an enumerator class for the DLPack device types has been added to `_usmarray.pyx`

This pull request also slips in a minor change to `usm_ndarray`'s constructor to honor the writable flag when `buffer` is another `usm_ndarray`

- [X] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

**Comment by ndgrigorian** [comment]
@oleksandr-pavlyk 
The commit from the memory work branch seems to have broken the dlpack tests on Windows.

File: dpctl/tensor/_dlpack.pyx [hunk 1]
```diff
@@ -296,26 +333,192 @@ cpdef to_dlpack_capsule(usm_ndarray usm_ary):
...
+    if copied:
+        dlmv_flags |= DLPACK_FLAG_BITMASK_IS_COPIED
+    if not (flags & USM_ARRAY_WRITABLE):
+        dlmv_flags |= DLPACK_FLAG_BITMASK_READ_ONLY
+    dlmv_tensor.flags = dlmv_flags
+
+    dlmv_tensor.version.major = DLPACK_MAJOR_VERSION
+    dlmv_tensor.version.minor = DLPACK_MINOR_VERSION
+
+    dlmv_tensor.manager_ctx = <void*>usm_ary
```
Comment (oleksandr-pavlyk) [inline dpctl/tensor/_dlpack.pyx:501]:
```suggestion
    dlmv_tensor.manager_ctx = <void*>ary_base
```

**Comment by oleksandr-pavlyk** [comment]
The check need tweaking:

```
In [9]: x.__dlpack__(max_version=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 1
----> 1 x.__dlpack__(max_version=1)

File ~/repos/dpctl_tmp/dpctl/tensor/_usmarray.pyx:1131, in dpctl.tensor._usmarray.usm_ndarray.__dlpack__()
   1129 else:
   1130     dpctl_dlpack_version = get_build_dlpack_version()
-> 1131     if max_version >= dpctl_dlpack_version or max_version[0] == dpctl_dlpack_version[0]:
   1132         # DLManagedTensorVersioned path
   1133         # TODO: add logic for targeting a device

TypeError: '>=' not supported between instances of 'int' and 'tuple'
```

The check should check that `max_version` is a 2-tuple of integers. But looking at this check, what is the purpose of overriding clause `max_version[0] == dpctl_dlpack_version[0]`?

**Comment by oleksandr-pavlyk** [comment]
Please update `dpctl/tensor/include/dlpack/README.md` with version and SHA of the vendored `dlpack.h`

**Comment by ndgrigorian** [comment]
> The check need tweaking:
> 
> ```
> In [9]: x.__dlpack__(max_version=1)
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> Cell In[9], line 1
> ----> 1 x.__dlpack__(max_version=1)
> 
> File ~/repos/dpctl_tmp/dpctl/tensor/_usmarray.pyx:1131, in dpctl.tensor._usmarray.usm_ndarray.__dlpack__()
>    1129 else:
>    1130     dpctl_dlpack_version = get_build_dlpack_version()
> -> 1131     if max_version >= dpctl_dlpack_version or max_version[0] == dpctl_dlpack_version[0]:
>    1132         # DLManagedTensorVersioned path
>    1133         # TODO: add logic for targeting a device
> 
> TypeError: '>=' not supported between instances of 'int' and 'tuple'
> ```
> 
> The check should check that `max_version` is a 2-tuple of integers. But looking at this check, what is the purpose of overriding clause `max_version[0] == dpctl_dlpack_version[0]`?

`max_version[0] >= dpctl_dlpack_version[0]` should be enough.

The spec recommends that if the major versions match, it's a strong enough condition to return a capsule with our own `max_version`.

**Comment by oleksandr-pavlyk** [comment]
It would be good to factor out common code from `to_dlpack_capsule` and `to_dlpack_versioned_capsule` into a function.

**Review by oleksandr-pavlyk — APPROVED** [review]
LGTM! Thank you @ndgrigorian

**Pull request merged**